src/graphics/opengl/CommandBufferGL.h

```diff
@@ -0,0 +1,135 @@
+// Copyright © 2008-2021 Pioneer Developers. See AUTHORS.txt for details
+// Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
+
+#pragma once
+
+#include "Color.h"
+#include "graphics/Graphics.h"
+#include "graphics/Types.h"
+#include "graphics/VertexBuffer.h"
+
+#include "OpenGLLibs.h"
+#include <variant>
+
+namespace Graphics {
+
+	class Material;
+	class MeshObject;
+	class VertexArray;
+
+	class RendererOGL;
+
+	namespace OGL {
+
+		class InstanceBuffer;
+		class Material;
+		class MeshObject;
+		class Program;
+		class Shader;
+		class TextureGL;
+		class RenderTarget;
+		class IndexBuffer;
+		class UniformBuffer;
+		class VertexBuffer;
+
+		class CommandList {
+		public:
+			struct DrawCmd {
+				MeshObject *mesh;
+				InstanceBuffer *inst = nullptr;
+				const Shader *shader = nullptr;
+				Program *program = nullptr;
+				size_t renderStateHash = 0;
+				char *drawData;
+			};
+
+			struct DynamicDrawCmd {
+				BufferBinding<VertexBuffer> vtxBind;
+				BufferBinding<IndexBuffer> idxBind;
+				const Shader *shader = nullptr;
+				Program *program = nullptr;
+				size_t renderStateHash = 0;
+				char *drawData;
+			};
+
+			struct RenderPassCmd {
+				RenderTarget *renderTarget;
+				ViewportExtents extents;
+				ViewportExtents scissor;
+				bool setRenderTarget;
+				bool setScissor;
+				bool clearColors;
+				bool clearDepth;
+				Color clearColor;
+			};
+
+			// development asserts to ensure sizes are kept reasonable.
+			// if you need to go beyond these sizes, add a new command instead.
+			static_assert(sizeof(DrawCmd) <= 64);
+			static_assert(sizeof(DynamicDrawCmd) <= 64);
+			static_assert(sizeof(RenderPassCmd) <= 64);
+
+			void AddDrawCmd(Graphics::MeshObject *mesh, Graphics::Material *mat, Graphics::InstanceBuffer *inst = nullptr);
+			void AddDynamicDrawCmd(BufferBinding<Graphics::VertexBuffer> vtx, BufferBinding<Graphics::IndexBuffer> idx, Graphics::Material *mat);
+
+			void AddRenderPassCmd(RenderTarget *renderTarget, ViewportExtents extents);
+			void AddScissorCmd(ViewportExtents extents);
+			void AddClearCmd(bool clearColors, bool clearDepth, Color color);
+
+		protected:
+			using Cmd = std::variant<DrawCmd, DynamicDrawCmd, RenderPassCmd>;
+			const std::vector<Cmd> &GetDrawCmds() const { return m_drawCmds; }
+
+			bool IsEmpty() const { return m_drawCmds.empty(); }
+			void Reset();
+
+		private:
+			friend class Graphics::RendererOGL;
+			CommandList(Graphics::RendererOGL *r) :
+				m_renderer(r)
+			{
+				m_drawCmds.reserve(32);
```

Comment:
I wonder why 32?

Reply:
STL vectors typically allocate in increasing powers of two. Reserving 8 elements is usually considered the "magic number" to prevent spurious allocations without also reserving too much memory, but generally speaking I expect *at least* 32 commands to be written to a command list (to make up for the cost of submitting the command list). This could honestly be bumped up higher, as all 700+ draw commands are recorded in the same command list, but I eventually want to separate the scene draw into multiple command lists so it stays at 32 for now.

Reply:
I suppose stuff like that can warrant a line or two as comment in the code.

Reply:
@impaktor imhi it doesn't seem to matter much, because after the very first frame the capacity of this vector will greatly increase and will no longer decrease, if I'm not mistaken